fmt/src/ast_eq.rs

```diff
@@ -0,0 +1,508 @@
+                        let left = ($($tuple_field),*);
+                        if let $name::$tuple_variant(.., $($tuple_field),*) = other {
+                            let right = ($($tuple_field),*);
+                            left.ast_eq(&right)
+                        } else {
+                            false
+                        }
+                    }
+                    )*
+                    $(
+                    $name::$struct_variant { $($struct_field),* , .. } => {
+                        let left = ($($struct_field),*);
+                        if let $name::$struct_variant { $($struct_field),* , .. } = other {
+                            let right = ($($struct_field),*);
+                            left.ast_eq(&right)
+                        } else {
+                            false
+                        }
+                    }
+                    )*
+                }
+            }
+        }
+    }
+}
+
+derive_ast_eq! { (0 A) }
+derive_ast_eq! { (0 A, 1 B) }
+derive_ast_eq! { (0 A, 1 B, 2 C) }
+derive_ast_eq! { (0 A, 1 B, 2 C, 3 D) }
+derive_ast_eq! { (0 A, 1 B, 2 C, 3 D, 4 E) }
+derive_ast_eq! { String }
+derive_ast_eq! { bool }
+derive_ast_eq! { u8 }
+derive_ast_eq! { u16 }
+derive_ast_eq! { struct Identifier { name } }
+derive_ast_eq! { struct HexLiteral { hex } }
+derive_ast_eq! { struct StringLiteral { string } }
+derive_ast_eq! { struct Parameter { ty, storage, name } }
+derive_ast_eq! { struct NamedArgument { name, expr } }
+derive_ast_eq! { struct YulBlock { statements } }
+derive_ast_eq! { struct DocComment { ty } }
+derive_ast_eq! { struct YulFunctionCall { id, arguments } }
+derive_ast_eq! { struct YulFunctionDefinition { id, params, returns, body } }
+derive_ast_eq! { struct YulSwitch { condition, cases, default } }
+derive_ast_eq! { struct YulFor {
+    init_block,
+    condition,
+    post_block,
+    execution_block,
+}}
+derive_ast_eq! { struct YulTypedIdentifier { id, ty } }
+derive_ast_eq! { struct VariableDeclaration { ty, storage, name } }
+derive_ast_eq! { struct Using { list, ty, global } }
+derive_ast_eq! { struct TypeDefinition { name, ty } }
+derive_ast_eq! { struct ContractDefinition { ty, name, base, parts } }
+derive_ast_eq! { struct EventParameter { ty, indexed, name } }
+derive_ast_eq! { struct ErrorParameter { ty, name } }
+derive_ast_eq! { struct EventDefinition { name, fields, anonymous } }
+derive_ast_eq! { struct ErrorDefinition { name, fields } }
+derive_ast_eq! { struct StructDefinition { name, fields } }
+derive_ast_eq! { struct EnumDefinition { name, values } }
+derive_ast_eq! { enum CommentType {
+    Line,
+    Block,
+    _
+    _
+}}
+derive_ast_eq! { enum UsingList {
+    _
+    Library(expr),
+    Functions(exprs),
+    _
+}}
+derive_ast_eq! { enum Visibility {
+    _
+    External(),
+    Public(),
+    Internal(),
+    Private(),
+    _
+}}
+derive_ast_eq! { enum Mutability {
+    _
+    Pure(),
+    View(),
+    Constant(),
+    Payable(),
```

Comment:
Should these be `Pure` instead of `Pure()`? Given it's an enum with no internal value? @seanyoung maybe a Solang thing